src/starkware_utils/commitment_tree/update_tree.rs

```diff
@@ -0,0 +1,375 @@
+use std::collections::{HashMap, HashSet};
+use std::marker::PhantomData;
+
+use num_bigint::BigUint;
+
+use crate::starkware_utils::commitment_tree::base_types::{Height, TreeIndex};
+use crate::starkware_utils::commitment_tree::binary_fact_tree::BinaryFactDict;
+use crate::starkware_utils::commitment_tree::binary_fact_tree_node::BinaryFactTreeNode;
+use crate::starkware_utils::commitment_tree::calculation::{Calculation, CalculationNode, NodeFactDict};
+use crate::starkware_utils::commitment_tree::errors::{TreeError, UpdateTreeError};
+use crate::starkware_utils::commitment_tree::inner_node_fact::InnerNodeFact;
+use crate::starkware_utils::commitment_tree::leaf_fact::LeafFact;
+use crate::starkware_utils::commitment_tree::merkle_tree::traverse_tree::{traverse_tree, TreeTraverser};
+use crate::starkware_utils::commitment_tree::patricia_tree::nodes::PatriciaNodeFact;
+use crate::storage::storage::{DbObject, FactFetchingContext, HashFunctionType, Storage, StorageError};
+
+#[derive(Clone, Debug)]
+enum TreeUpdate<LF>
+where
+    LF: Clone,
+{
+    Tuple(Box<UpdateTree<LF>>, Box<UpdateTree<LF>>),
+    Leaf(LF),
+}
+
+type UpdateTree<LF> = Option<TreeUpdate<LF>>;
+
+/// Checks if there are merkle nodes that are not updated, but all of its children are.
+/// Starts at node_index, and goes up to the root.
+async fn update_necessary<S, H, LF, CN>(
+    node_index: TreeIndex,
+    updated_nodes: &mut HashMap<TreeIndex, CN>,
+    ffc: &mut FactFetchingContext<S, H>,
+    facts: &mut Option<BinaryFactDict>,
+) -> Result<(), UpdateTreeError>
+where
+    S: Storage + Sync + Send,
+    H: HashFunctionType + Sync + Send,
+    LF: LeafFact<S, H>,
+    CN: CalculationNode<S, H, LF>,
+{
+    let mut cur_node_index = node_index;
+
+    // Xoring by 1 switch between 2k <-> 2k + 1, which are siblings in the tree.
+    // The parent of these siblings is k = floor(n/2) for n = 2k, 2k+1.
+    while updated_nodes.contains_key(&(cur_node_index.clone() ^ BigUint::from(1u64))) {
```

Comment:
```suggestion
    let one = BigUint::from(1u64);
    while updated_nodes.contains_key(&(cur_node_index.clone() ^ one)) {
```

might be minor, but calling `from()` repeatedly looks like it incurs a heap allocation each time

Reply:
At first sight I thought you were right, I tried it but BigUint is not `Copy` so I'd have to clone it each time anyway. Not sure whether the compiler generates different code for multiple calls to `from()` and `clone()` :man_shrugging:

Reply:
You can compare by reference, pass `&one` instead of `one`. Nice!

Reply:
Good point, that works!